code/__HELPERS/filters.dm

```diff
@@ -0,0 +1,319 @@
+			"color" = matrix(),
+			"space" = FILTER_COLOR_RGB
+		)
+	),
+	*/
+	"displace" = list(
+		"defaults" = list(
+			"x" = 0,
+			"y" = 0,
+			"size" = null,
+			"icon" = ICON_NOT_SET,
+			"render_source" = ""
+		)
+	),
+	"drop_shadow" = list(
+		"defaults" = list(
+			"x" = 1,
+			"y" = -1,
+			"size" = 1,
+			"offset" = 0,
+			"color" = COLOR_HALF_TRANSPARENT_BLACK
+		)
+	),
+	"blur" = list(
+		"defaults" = list(
+			"size" = 1
+		)
+	),
+	"layer" = list(
+		"defaults" = list(
+			"x" = 0,
+			"y" = 0,
+			"icon" = ICON_NOT_SET,
+			"render_source" = "",
+			"flags" = FILTER_OVERLAY,
+			"color" = "",
+			"transform" = null,
+			"blend_mode" = BLEND_DEFAULT
+		)
+	),
+	"motion_blur" = list(
+		"defaults" = list(
+			"x" = 0,
+			"y" = 0
+		)
+	),
+	"outline" = list(
+		"defaults" = list(
+			"size" = 0,
+			"color" = COLOR_BLACK,
+			"flags" = NONE
+		),
+		"flags" = list(
+			"OUTLINE_SHARP" = OUTLINE_SHARP,
+			"OUTLINE_SQUARE" = OUTLINE_SQUARE
+		)
+	),
+	"radial_blur" = list(
+		"defaults" = list(
+			"x" = 0,
+			"y" = 0,
+			"size" = 0.01
+		)
+	),
+	"rays" = list(
+		"defaults" = list(
+			"x" = 0,
+			"y" = 0,
+			"size" = 16,
+			"color" = COLOR_WHITE,
+			"offset" = 0,
+			"density" = 10,
+			"threshold" = 0.5,
+			"factor" = 0,
+			"flags" = FILTER_OVERLAY | FILTER_UNDERLAY
+		),
+		"flags" = list(
+			"FILTER_OVERLAY" = FILTER_OVERLAY,
+			"FILTER_UNDERLAY" = FILTER_UNDERLAY
+		)
+	),
+	"ripple" = list(
+		"defaults" = list(
+			"x" = 0,
+			"y" = 0,
+			"size" = 1,
+			"repeat" = 2,
+			"radius" = 0,
+			"falloff" = 1,
+			"flags" = NONE
+		),
+		"flags" = list(
+			"WAVE_BOUNDED" = WAVE_BOUNDED
+		)
+	),
+	"wave" = list(
+		"defaults" = list(
+			"x" = 0,
+			"y" = 0,
+			"size" = 1,
+			"offset" = 0,
+			"flags" = NONE
+		),
+		"flags" = list(
+			"WAVE_SIDEWAYS" = WAVE_SIDEWAYS,
+			"WAVE_BOUNDED" = WAVE_BOUNDED
+		)
+	)
+))
+
+#undef ICON_NOT_SET
+
+//Helpers to generate lists for filter helpers
+//This is the only practical way of writing these that actually produces sane lists
+/proc/alpha_mask_filter(x, y, icon/icon, render_source, flags)
+	. = list("type" = "alpha")
+	if(!isnull(x))
+		.["x"] = x
+	if(!isnull(y))
+		.["y"] = y
+	if(!isnull(icon))
+		.["icon"] = icon
+	if(!isnull(render_source))
+		.["render_source"] = render_source
+	if(!isnull(flags))
+		.["flags"] = flags
+
+/proc/angular_blur_filter(x, y, size)
+	. = list("type" = "angular_blur")
+	if(!isnull(x))
+		.["x"] = x
+	if(!isnull(y))
+		.["y"] = y
+	if(!isnull(size))
+		.["size"] = size
+
+/proc/color_matrix_filter(matrix/in_matrix, space)
+	. = list("type" = "color")
+	.["color"] = in_matrix
+	if(!isnull(space))
+		.["space"] = space
+
+/proc/displacement_map_filter(icon, render_source, x, y, size = 32)
+	. = list("type" = "displace")
+	if(!isnull(icon))
+		.["icon"] = icon
+	if(!isnull(render_source))
+		.["render_source"] = render_source
+	if(!isnull(x))
+		.["x"] = x
+	if(!isnull(y))
+		.["y"] = y
+	if(!isnull(size))
+		.["size"] = size
```

Comment:
Please explain the logic of this to me, it seems like an utter waste

Reply:
This is the only way of doing this.

Reply:
that's a real shame, but you're right.